pcsc/src/lib.rs

```diff
@@ -1173,6 +1173,31 @@ impl Card {
         }
     }
 
+    /// Like `Card::transaction`, but returns the reference to `self` on error.
+    ///
+    /// This function is like [`Card::transaction`], but also returns
+    /// the reference to `self` on error.  When starting a
+    /// transaction, it is necessary to deal with transient errors,
+    /// like [`Error::ResetCard`] by reconnecting to the card, and
+    /// retrying the transaction.  When this functionality is wrapped,
+    /// this doesn't work, because mutable references can't be
+    /// reborrowed.  This function returns the reference, which allows
+    /// this construct.
+    pub fn transaction2(
+        &mut self,
+    ) -> Result<Transaction, (&mut Self, Error)> {
+        // We can't use try_pcsc, because it returns an incompatible
+        // type.  And we can't wrap it in a closure due to lifetimes.
+        // Hence we inline try_pcsc!.
```

Comment:
@nwalfield just for your information, as it may interest you, this is actually possible to use a closure in this case (but it ends up with a version which is much less readable unfortunately).

The biggest problem is that `map` expect a `FnOnce`, so in order to convert the error type, we cannot capture `self` as a reference as `FnOnce` captures all by move.

The workaround is to either use raw pointer (we already are in an unsafe block):

``` rust
    pub fn transaction2(
        &mut self,
    ) -> Result<Transaction, (&mut Self, Error)> {
            unsafe {
                let this : *mut Self = self;
                (||{
                    Ok(try_pcsc!(ffi::SCardBeginTransaction(
                        self.handle,
                    )))
                })()
                .map(|_| Transaction{card: &mut *this})
                .map_err(|e| (&mut *this, e))
            }
    }
```

Or use a `match` to avoid to use the `map` function call:

``` rust
    pub fn transaction2(
        &mut self,
    ) -> Result<Transaction, (&mut Self, Error)> {
            unsafe {
                match (||{
                    Ok(try_pcsc!(ffi::SCardBeginTransaction(
                        self.handle,
                    )))
                })()
                {
                    Ok(_) => Ok(Transaction{card: self}),
                    Err(e) => Err((self, e)),
                }
            }
    }
```

Wrapping `try_pcsc` in a closure is not really an issue as this closure as no specific constraints and compiler can choose to capture variables by reference (making a `Fn` essentially).

But as I said, this is just done by curiosity, the final result is not very appealing ^^

Reply:
Thanks for sharing!  I tried using a closure, but it didn't work for me.  I thought I did what you suggest in your second workaround, but I must have messed something up.